modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2TokenExchangeGrantTypeTestCase.java

```diff
@@ -157,30 +176,238 @@ public void endTest() throws Exception {
+
+        //assert local subject identifier config is mandatory, implicit association config disabled
+        updateIdentityProviderAssociationConfig(false);
+        updateAssertLocalSubjectIdentifierConfig(IdentityConstants.AssertLocalSubjectMode.MANDATORY);
+        List<NameValuePair> postParameters = getTokenExchangePostParameters();
+        AssertUnauthorizedResponse(postParameters);
+    }
+
+    @Test(groups = "wso2.is", description = "Exchange access token for local user with no local account",
+            dependsOnMethods = "testTokenExchangeForLocalUserWithImplicitAssociationConfigDisabled")
+    public void testTokenExchangeForLocalUserWithNoLocalAccount() throws Exception {
+
+        //assert local subject identifier config is mandatory, no matching local user account
+        deleteUser(PORT_OFFSET_0, NEW_USER_USERNAME);
+        List<NameValuePair> postParameters = getTokenExchangePostParameters();
+        AssertUnauthorizedResponse(postParameters);
+    }
+
+    private List<NameValuePair> getTokenExchangePostParameters() {
 
         List<NameValuePair> postParameters = new ArrayList<>();
         postParameters.add(new BasicNameValuePair("subject_token", accessTokenFromSecondaryIS));
         postParameters.add(new BasicNameValuePair("subject_token_type", "urn:ietf:params:oauth:token-type:jwt"));
         postParameters.add(new BasicNameValuePair("requested_token_type", "urn:ietf:params:oauth:token-type:jwt"));
         postParameters.add(new BasicNameValuePair("grant_type", "urn:ietf:params:oauth:grant-type:token-exchange"));
+        return postParameters;
+    }
+
+    private void AssertExchangedTokenForFederatedUser(List<NameValuePair> postParameters) throws Exception {
+
         JSONObject responseObject = sendPOSTMessage(PRIMARY_IS_TOKEN_ENDPOINT, primaryISClientID, primaryISClientSecret,
                 postParameters);
         String exchangedToken = responseObject.get("access_token").toString();
+        Assert.assertNotNull(exchangedToken, "Access token is null.");
+        Assert.assertEquals(getTokenSubject(exchangedToken), secondaryISUserId, "Subject of the exchanged " +
+                "token should be the same as the subject of the subject token.");
+    }
 
+    private void AssertExchangedTokenForLocalUser(List<NameValuePair> postParameters) throws Exception {
+
+        JSONObject responseObject = sendPOSTMessage(PRIMARY_IS_TOKEN_ENDPOINT, primaryISClientID, primaryISClientSecret,
+                postParameters);
+        String exchangedToken = responseObject.get("access_token").toString();
         Assert.assertNotNull(exchangedToken, "Access token is null.");
+        Assert.assertEquals(getTokenSubject(exchangedToken), primaryISUserId, "Subject of the exchanged " +
+                "token should be the same as the subject of the matching local user.");
+    }
+
+    private void AssertUnauthorizedResponse(List<NameValuePair> postParameters) throws Exception {
```

Comment:
```suggestion
    private void assertUnauthorizedResponse(List<NameValuePair> postParameters) throws Exception {
```

Reply:
addressed with [75ace20](https://github.com/wso2/product-is/pull/17605/commits/75ace2037227dfc5018c6c9afff5b3e83995533d)